activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -9,8 +9,7 @@ def handle_dependency
 
         case options[:dependent]
         when :destroy
-          target.destroy
-          raise ActiveRecord::Rollback unless target.destroyed?
+          raise ActiveRecord::Rollback unless target.destroy
```

Comment:
when `ActiveRecord::Rollback` is raised the result of `target.destroy` is `nil`. Relying on that allows to handle nested `belongs_to` with `dependent: :destroy` correctly. 
Previously `destroyed?` relied on `@destroyed` which is `true` with callbacks nested starting three or more levels.
This change also ensures that `destroyed?` returns correctly `false` in this case.

Reply:
@intrip nice, thank you for the explanation and elegant solution. I remember that I have tried to get somehow transaction status, but gave up.